### PR TITLE
Feat/setup talk2cells

### PR DESCRIPTION
**Pull request by gurdeep330**
# For authors

https://github.com/user-attachments/assets/820510e7-d5b2-40af-a7ce-fe05b268c62f

### Talk2Cells Agent Initial Implementation

This PR introduces the initial implementation of the `talk2cells` agentic workflow using Langgraph. The current implementation defines a graph with a single node, the **Single-Cell Protocol Agent** (`scp_agent`), and includes two tools. The graph has a state containing one key: `search_table`.

### 1. `search_studies`
This tool searches for studies based on specific search terms in the Single-Cell Protocol database. It demonstrates how a tool can modify the graph's state.  

- **Functionality**:  
  Upon retrieving results, the tool updates the graph's state by writing the results to the key `search_table`.  
- **Implementation**:  
  The tool uses the `Command` object in its return statement to perform this state update.

### 2. `display_studies`
This tool reads the contents of the state (key `search_table`) and displays the data back to the user on the front end. It serves as an example of how to share the graph's state with another tool.  

- **Functionality**:  
  The state is passed as an argument using Langgraph's prebuilt module `InjectedState`. This argument is hidden from the LLM.  
- **Implementation**:  
  The tool accesses the existing state, processes it, and delivers the data for display without altering the state.

#### Additional Notes:
- More agents will be added to this workflow in the coming weeks.
- This workflow has its known streamlit app for now. Once the other agents have also migrated to Lnaggraph, their streamlit apps will merged into a single app and led by a main agent.
- Currently, the system prompt for Talk2Cells is defined within the `scp_agent`, but this will eventually handled by Hydra in a [different issue](https://github.com/VirtualPatientEngine/AIAgents4Pharma/issues/61#issue-2795733638).

#### Revision:
- Logging and Langsmith tracer were added for observability.

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [ ] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers
## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: app/frontend/streamlit_app_talk2cells.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                {
+                    "type": "message",
+                    "content": prompt_msg
+                }
+            )
+            with st.chat_message("user", avatar="👩🏻‍💻"):
+                st.markdown(prompt)
+                st.empty()
+
+            with st.chat_message("assistant", avatar="🤖"):
```
Comment (dmccloskey) [inline app/frontend/streamlit_app_talk2cells.py:113]: the langsmith client is missing for feedback. I assume this will also come in a subsequent PR?

File: aiagents4pharma/talk2cells/tests/scp_agent/test_scp_agent.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
+'''
+Test cases for the search_studies
+'''
+
+# from ..tools.search_studies import search_studies
+from aiagents4pharma.talk2cells.agents.scp_agent import get_app
+from langchain_core.messages import HumanMessage
+
+def test_agent_scp():
```
Comment (dmccloskey) [inline aiagents4pharma/talk2cells/tests/scp_agent/test_scp_agent.py:9]: the `traceable` decorators are missing. I assume observability will come in a subsequent PR?

**Review by dmccloskey — APPROVED** [review]
Nice work 💪. I don't have any major comments on the code. A major concern was already addressed by the new issue of using Hydra for configs. Another concern is the addition of langsmith and observability. I would recommend having this added in early on. If you agree, I would recommend adding another issue for observability.

**Comment by gurdeep330** [comment]
Thanks for the comments, @dmccloskey. Initially, I was planning to add tracers when we connect all the agents later. However, I think having separate tracers for each agent makes sense (we can always refactor them later/still have one for the main agent). I have updated the PR:
- Added logging (to the agents and tools code) and tracer (to the Streamlit code)
- Updated PR description
- Updated video in the PR

File: app/frontend/streamlit_app_talk2cells.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+                    # # Get response from the agent
+                    # response = app.invoke(
+                    #     {"messages": [HumanMessage(content=prompt)]},
+                    #     config=config
+                    # )
+                    ERROR_FLAG = False
+                    with collect_runs() as cb:
+                        # Add Langsmith tracer
+                        tracer = LangChainTracer(
```
Comment (dmccloskey) [inline app/frontend/streamlit_app_talk2cells.py:174]: 👍

File: app/frontend/streamlit_app_talk2cells.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+# Initialize graph
+if "unique_id" not in st.session_state:
+    st.session_state.unique_id = random.randint(1, 1000)
+if "app" not in st.session_state:
+    st.session_state.app = get_app(st.session_state.unique_id)
+
+# Get the app
+app = st.session_state.app
+
+def _submit_feedback(user_response):
```
Comment (dmccloskey) [inline app/frontend/streamlit_app_talk2cells.py:59]: 👍

**Review by dmccloskey — APPROVED** [review]
Looks nice 🤩. 

The addition application logging, Tracing, and Feedback will provide a solid template for the other agents and a good working example for onboarding potential contributors
